.Gruntfile.babel.js

```diff
@@ -243,6 +268,26 @@ module.exports = (grunt) => {
     }
   })
 
+  grunt.registerTask('backend:launch', function () {
+    const done = this.async()
+    grunt.log.writeln('backend: launching...')
+
+    // I don't understand how test:e2e works but `grun test` doesn't...
+    // this doesn't seem to start the server, but somehow it starts...
+    child = fork('Gruntfile.js', process.argv, {
+      env: { LOAD_NO_FILE: true, ...process.env }
+    })
+
+    child.on('exit', (c) => {
+      if (c !== 0) {
+        grunt.log.error(`child exited with error code: ${c}`.bold)
+        // ^C can cause c to be null, which is an OK error
+        process.exit(c || 0)
+      }
+    })
+    done()
+  })
+
```

Comment:
It's not clear to me why you've created this function... The file `test/backend.js` should be moved into the cypress tests instead.

Reply:
I think you didn't read my previous comments, please take a look at them: https://github.com/okTurtles/group-income-simple/pull/625#discussion_r317085213

Reply:
@greg Please read this:

![image](https://user-images.githubusercontent.com/14869087/63650309-2dff0480-c741-11e9-9e96-8e65b982eea7.png)

This screenshot appears when we run cypress without having a server running.

One of the best practices of Cypress is: 

> Anti-Pattern: Don’t try to start a web server from within Cypress scripts. Read about [best practices](https://docs.cypress.io/guides/references/best-practices.html#Web-Servers) here.
> ( ... )
> Best Practice: Start a web server prior to running Cypress in the Test Runner or headless mode.

You asking to add the test "Should start the server" to cypress is not correct for many reasons (read best practices). Plus, we already know that the server starts correctly because we tested it on the unit tests (`tests/backend.js file`).

So, we need to start the server before the Cypress task. That's the purpose of `grunt:launch` task.

```js
    grunt.registerTask('test:e2e', ['dist', 'connect', 'backend:launch', 'cypress'])
    grunt.registerTask('test', ['dist', 'connect', 'exe:test', 'backend:launch', 'cypress'])
```

I hope you understand now. Can you teach me how to start the server correctly (and how to shut it down to avoid any kind of memoryleak/port conflicts?

Reply:
I did read over your comments. We are talking about different servers here. The error you're seeing for port 8000 should not be an issue since that is started by the `connect` task. I think you should still start the backend server (which runs on port 3000) in Cypress. However, I think at this point it's probably best for me to try to do this myself so that maybe I'll see what errors you're running into. It's possible that my recommendations are off for some reason. I'll give this a try today and will let you know what I find!

> Can you teach me how to start the server correctly and how to shut it down when cypress finishes, so we avoid any kind of memory leak/port conflicts?

Yes, I believe you should start the 3000 server as it used to be started in both `test/backend.js` and `test/frontend.js` - no different. As for shutting it down, I don't think you need to do that, but if you want you can have a look at how the Gruntfile.babel.js does it, and also look into `backend/index.js`.

Reply:
Regarding the Cypress documentation, in the link you posted, here's what it says:

>  Anti-Pattern: Trying to a start a web server from within Cypress scripts with `cy.exec()` or `cy.task()`.

Perhaps I'm missing something but I don't think you would be using either of those functions to start the 3000 server, unless Cypress makes it impossible to do it otherwise, which I wouldn't know because I haven't given it a try myself yet, so that's what I should do to find out. I'll report back to you!